script/zonemaster_backend_rpcapi.psgi

```diff
@@ -66,7 +66,10 @@ builder {
         my $app = shift;
 
         # Make sure we can connect to the database
-        $config->new_DB();
+        my $db = $config->new_DB();
+
+        $log->notice( "Initializing the database" );
+        $db->create_db();
```

Comment:
Could we detect if the database is initialized and only initialize it if it isn't? I don't think it's helpful to have log lines that claim that the database was initialized when in fact nothing was changed. The same goes for zonemaster_backend_testagent. Also there seems to be a race condition where multiple daemons compete to perform the same initialization.

Reply:
I don't think we currently can. And I agree that this would be better to have this kind of detection. This also goes with your question about whether I have a plan about database migrations. Since database migration is about running a Perl script, I would like to have them being run automatically after an update. This would require to be able to check the version of the previous installed Backend and compare it to the newly installed version. Then we can automatically run the corresponding upgrade routines. However I haven't found a solution on how to compare these versions using our current code yet. Maybe we need to store it in database so that we can perform this comparison.
If we have a way to compare versions, then we will be able to detect if the database is initialized or not.

About the race condition, I have to review the code, but I think that all clauses using `IF NOT EXISTS` should not be affected.

Reply:
> Maybe we need to store it in database so that we can perform this comparison. If we have a way to compare versions, then we will be able to detect if the database is initialized or not.

I definitely think that we should store the database structure version in the database.

Reply:
> Also there seems to be a race condition where multiple daemons compete to perform the same initialization.
>
> > About the race condition, I have to review the code, but I think that all clauses using IF NOT EXISTS should not be affected.

Actually my point is more about the architecture than the implementation.

All daemons will attempt to perform every part of the initialization. Not only in this version but in every future version. Since I think we agree that schema initialization should be performed using the same approach, it needs to work for upgrades too.

The fewer failure modes I have to worry about during schema upgrades (and initialization too) the happier I am. I would really prefer to have these operations be performed in a friendlier single threaded context.